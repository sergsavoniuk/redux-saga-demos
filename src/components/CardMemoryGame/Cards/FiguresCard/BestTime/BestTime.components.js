import styled from 'styled-components';

export const Wrapper = styled.div`
  display: flex;
  flex-direction: column;
  width: 50%;
`;

export const Row = styled.div`
  display: flex;
  justify-content: space-between;
  padding: 5px;
`;

export const Label = styled.span``;
export const Value = styled.span``;
