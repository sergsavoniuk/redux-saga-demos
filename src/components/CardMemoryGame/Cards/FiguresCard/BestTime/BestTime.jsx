import React from 'react';
import { format } from 'date-fns';

import {
  Wrapper,
  Row,
  Span as Label,
  Span as Value,
} from './BestTime.components';

const DEFAULT_PLACEHOLDER = '-:-';
const TIME_FORMAT = 'ss.SSS';

function BestTime({ casual, medium, hard }) {
  return (
    <Wrapper>
      <Row>
        <Label>Best Casual:</Label>
        <Value>
          {casual ? `${format(casual, TIME_FORMAT)}s` : DEFAULT_PLACEHOLDER}
        </Value>
      </Row>
      <Row>
        <Label>Best Medium:</Label>
        <Value>
          {medium ? `${format(medium, TIME_FORMAT)}s` : DEFAULT_PLACEHOLDER}
        </Value>
      </Row>
      <Row>
        <Label>Best Hard:</Label>
        <Value>
          {hard ? `${format(hard, TIME_FORMAT)}s` : DEFAULT_PLACEHOLDER}
        </Value>
      </Row>
    </Wrapper>
  );
}

export default BestTime;
