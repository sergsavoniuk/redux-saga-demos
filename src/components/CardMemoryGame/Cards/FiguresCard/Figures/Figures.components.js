import styled from 'styled-components';

export const Title = styled.h2`
  width: 100%;
  text-transform: uppercase;
`;

export const Won = styled.span`
  margin: 0 5px;
`;
export const Lost = styled.span``;

export const Label = styled.i`
  vertical-align: top;
  font-size: 0.6em;
  color: #ffffff66;
  text-transform: uppercase;
`;
export const Value = styled.b``;
