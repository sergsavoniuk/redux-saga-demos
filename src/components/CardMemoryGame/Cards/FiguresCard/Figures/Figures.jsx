import React from 'react';

import {
  Title,
  Won,
  Lost,
  Abandoned,
  Label,
  Value,
} from './Figures.components';

function Figures({ won, lost, abandoned }) {
  return (
    <>
      <Title>
        Figures:
        <Won>
          <Value>{won}</Value>
          <Label>won</Label>
        </Won>
        <Lost>
          <Value>{lost}</Value>
          <Label>lost</Label>
        </Lost>
        <Abandoned>
          <Value>{abandoned}</Value>
          <Label>abandoned</Label>
        </Abandoned>
      </Title>
    </>
  );
}

export default Figures;
