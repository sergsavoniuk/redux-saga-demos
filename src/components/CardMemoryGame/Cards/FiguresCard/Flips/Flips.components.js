export { Wrapper, Row, Span } from '../BestTime/BestTime.components';
