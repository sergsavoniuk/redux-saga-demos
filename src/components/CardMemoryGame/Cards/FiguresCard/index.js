export { default, FiguresCard } from './FiguresCard';
