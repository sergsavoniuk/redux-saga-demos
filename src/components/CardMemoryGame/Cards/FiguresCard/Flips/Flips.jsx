import React from 'react';

import { Wrapper, Row, Span as Label, Span as Value } from './Flips.components';

function Flips({ matched, wrong }) {
  return (
    <Wrapper>
      <Row>
        <Label>Total Flips:</Label>
        <Value>{2 * (matched + wrong)}</Value>
      </Row>
      <Row>
        <Label>Matched Flips:</Label>
        <Value>{matched}</Value>
      </Row>
      <Row>
        <Label>Wrong Flips:</Label>
        <Value>{wrong}</Value>
      </Row>
    </Wrapper>
  );
}

export default Flips;
