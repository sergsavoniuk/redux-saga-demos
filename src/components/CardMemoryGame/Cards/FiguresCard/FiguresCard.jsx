import React from 'react';
import { connect } from 'react-redux';

import withMemo from 'utils/withMemo';
import Figures from './Figures';
import BestTime from './BestTime';
import Flips from './Flips';
import { Card } from '../Cards.components';
import { Content, Meta } from './FiguresCard.components';
import { Selectors } from 'redux/cardGame/';

export function FiguresCard({
  name,
  isFlipped,
  statistics: {
    won,
    lost,
    abandoned,
    casualBestTime,
    mediumBestTime,
    hardBestTime,
    matchedFlips,
    wrongFlips,
  },
  onCardClick,
}) {
  return !isFlipped ? (
    <Card onClick={onCardClick}>{name}</Card>
  ) : (
    <Card flipped={isFlipped} onClick={onCardClick}>
      <Content>
        <Figures won={won} lost={lost} abandoned={abandoned} />
        <Meta>
          <BestTime
            casual={casualBestTime}
            medium={mediumBestTime}
            hard={hardBestTime}
          />
          <Flips matched={matchedFlips} wrong={wrongFlips} />
        </Meta>
      </Content>
    </Card>
  );
}

function mapStateToProps(state) {
  return {
    statistics: Selectors.getStatistics(state),
  };
}

const MemoizedFiguresCard = withMemo(FiguresCard, ['isFlipped']);

export default connect(mapStateToProps)(MemoizedFiguresCard);
