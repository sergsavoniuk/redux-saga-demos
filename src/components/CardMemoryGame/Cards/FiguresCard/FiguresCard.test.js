import React from 'react';
import { shallow } from 'enzyme';

import { FiguresCard } from './FiguresCard';
import Figures from './Figures';
import BestTime from './BestTime';
import Flips from './Flips';

describe('<FiguresCard />', () => {
  const onCardClick = jest.fn();

  function setup(props) {
    const defaultProps = {
      name: 'F',
      isFlipped: false,
      statistics: {},
      onCardClick,
    };

    const wrapper = shallow(<FiguresCard {...defaultProps} {...props} />);

    return {
      wrapper,
      props: defaultProps,
    };
  }

  afterEach(() => {
    onCardClick.mockClear();
  });

  test('renders correctly unflipped card', () => {
    const { wrapper, props } = setup();

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.text()).toMatch(props.name);
  });

  test('renders correctly flipped card', () => {
    const { wrapper } = setup({ isFlipped: true });

    expect(wrapper).toMatchSnapshot();
    expect(
      wrapper.containsAllMatchingElements([
        <Figures />,
        <BestTime />,
        <Flips />,
      ]),
    ).toEqual(true);
  });

  test('should flip a card after a click', () => {
    const { wrapper, props } = setup();

    wrapper.simulate('click');
    expect(onCardClick).toHaveBeenCalledTimes(1);

    wrapper.setProps({ isFlipped: true });
    expect(
      wrapper.containsAllMatchingElements([
        <Figures />,
        <BestTime />,
        <Flips />,
      ]),
    ).toEqual(true);

    wrapper.simulate('click');
    expect(onCardClick).toHaveBeenCalledTimes(2);

    wrapper.setProps({ isFlipped: false });
    expect(
      wrapper.containsAllMatchingElements([
        <Figures />,
        <BestTime />,
        <Flips />,
      ]),
    ).toEqual(false);
    expect(wrapper.text()).toMatch(props.name);
  });
});
