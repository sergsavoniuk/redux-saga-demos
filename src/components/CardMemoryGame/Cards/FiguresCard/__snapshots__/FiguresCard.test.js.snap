// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FiguresCard /> renders correctly flipped card 1`] = `
<styled.div
  flipped={true}
  onClick={[MockFunction]}
>
  <styled.div>
    <Figures />
    <styled.div>
      <BestTime />
      <Flips />
    </styled.div>
  </styled.div>
</styled.div>
`;

exports[`<FiguresCard /> renders correctly unflipped card 1`] = `
<styled.div
  onClick={[MockFunction]}
>
  F
</styled.div>
`;
